### PR TITLE
PADV-504: Implement LTI 1.3 reusable configuration

### PR DESCRIPTION
**Pull request by kuipumu**
## Ticket

https://agile-jira.pearson.com/browse/PADV-504

## Description

This PR adds compatibility for LTI 1.3 external configurations to the LTI consumer XBlock. This change allow the XBlock to use values from the external LTI configuration when enabled and setup. This PR also includes a fix that adds validation to the external_config field on the LTI consumer XBlock and also improves the behavior of the JS that handles the visibility of configuration fields on the LTI consumer XBlock.

- [x] Modify get_lti_1p3_launch_info on api module.
- [x] Modify validate_field_data on lti_xblock module.
- [x] Modify get_lti_advantage_ags_mode on models module.
- [x] Modify get_lti_advantage_deep_linking_enabled on models module.
- [x] Modify get_lti_advantage_deep_linking_launch_url on models module.
- [x] Modify get_lti_advantage_nrps_enabled on models module.
- [x] Modify _get_lti_1p3_consumer on models module.
- [x] Modify xblock_studio_view.js.
- [x] Include unit tests for all added and modified code.

## Testing

1. Install and configure xblock-lti-consumer
2. Install and configure openedx-ltistore (follow the setup steps from: https://github.com/openedx/xblock-lti-consumer/pull/239)
3. Create an IMS reference tool: https://lti-ri.imsglobal.org/lti/tools
4. Go to http://localhost:18000/admin/lti_store/externallticonfiguration/ and create a new external LTI 1.3 configuration.
5. Fill the "LTI 1.3 Private Key",  "LTI 1.3 Client ID" and "LTI 1.3 Tool Keyset URL" or "LTI 1.3 Tool Public Key" field.
6. Create a course and add a LTI 1.3 consumer XBlock.
7. Set the consumer XBlock to user an external configuration and set the reusable configuration ID (follow the steps on how to get the filter key from: https://github.com/openedx/xblock-lti-consumer/pull/239) 
8. Copy the keyset and access token URL from the XBlock preview back to the IMS reference tool previously created. 
9. Go to the course a launch the XBlock.
10. The LTI 1.3 launch should be valid.
11. Repeat the test while changing the deployment ID on the external LTI configuration and IMS reference tool.
12. Repeat the test setting the OIDC and launch URL on the XBlock and check if the values from the XBlock are being used instead of the values on the external configuration.
13. Check that the XBlock doesn't allow you to set the external configuration type without adding an external configuration ID.

## Reviewers

- [x] @squirrel18
- [x] @alexjmpb

### CODE REVIEW

File: lti_consumer/models.py [hunk 1]
```diff
@@ -625,6 +653,24 @@ class LtiAgsLineItem(models.Model):
         blank=True
     )
 
+    # OIDC URL, Client ID and Deployment ID
+    # This allows us to isolate the LineItem per tool deployment
+    oidc_url = models.CharField(
+        max_length=255,
+        blank=True,
+        null=True
```
Comment (alexjmpb) [inline lti_consumer/models.py]: Trailing comma

File: lti_consumer/models.py [hunk 2]
```diff
@@ -625,6 +653,24 @@ class LtiAgsLineItem(models.Model):
...
+    )
+    client_id = models.CharField(
+        max_length=255,
+        blank=True,
+        null=True
```
Comment (alexjmpb) [inline lti_consumer/models.py]: Trailing comma

File: lti_consumer/models.py [hunk 3]
```diff
@@ -625,6 +653,24 @@ class LtiAgsLineItem(models.Model):
...
+    )
+    deployment_id = models.CharField(
+        max_length=255,
+        blank=True,
+        null=True
```
Comment (alexjmpb) [inline lti_consumer/models.py]: Trailing comma

File: lti_consumer/plugin/views.py [hunk 4]
```diff
@@ -556,19 +556,25 @@ class LtiAgsLineItemViewset(viewsets.ModelViewSet):
...
-        # TODO:
-        # Note that each configuration currently maps 1:1
-        # to each resource link (block), and this filter needs
-        # improved once we start reusing LTI configurations.
         return LtiAgsLineItem.objects.filter(
-            lti_configuration=lti_configuration
+            lti_configuration=lti_configuration,
+            oidc_url=consumer_dict.get('oidc_url'),
+            client_id=consumer_dict.get('client_id'),
+            deployment_id=consumer_dict.get('deployment_id'),
```
Comment (alexjmpb) [inline lti_consumer/plugin/views.py:565]: Can these values be None, if so, would it affect in something?

Comment (kuipumu) [reply lti_consumer/plugin/views.py:565]: @alexjmpb They already return None if they don't exist

**Review by alexjmpb — APPROVED** [review]

File: lti_consumer/api.py [hunk 5]
```diff
@@ -141,14 +141,29 @@ def get_lti_1p3_launch_info(
             deep_linking_content_items = [item.attributes for item in dl_content_items]
 
     config_id = lti_config.config_id
+    client_id = lti_config.lti_1p3_client_id
+    token_url = get_lms_lti_access_token_link(config_id)
+    keyset_url = get_lms_lti_keyset_link(config_id)
+    deployment_id = '1'
```
Comment (Squirrel18) [inline lti_consumer/api.py:150]: Please, add a doc-string about this number. @kuipumu

File: lti_consumer/models.py [hunk 6]
```diff
@@ -441,7 +429,29 @@ def _setup_lti_1p3_ags(self, consumer):
             log.info('LTI Advantage AGS is disabled for %s', self)
             return
 
-        lineitem = self.ltiagslineitem_set.first()
+        deployment_id = 1
```
Comment (Squirrel18) [inline lti_consumer/models.py:435]: Please, add a doc-string about this hardcoded value. @kuipumu

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

**Review by Squirrel18 — APPROVED** [review]

**Pull request merged**